staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go

```diff
@@ -0,0 +1,252 @@
+		contextMapper: contextMapper,
+	}
+
+	// Get Local delegate's Spec
+	s.localDelegatesOpenAPISpec, err = s.loadOpenAPISpec(delegateHandler)
+	if err != nil {
+		return nil, err
+	}
+
+	// Build Aggregator's spec
+	s.aggregatorOpenAPISpec, err = builder.BuildOpenAPISpec(
+		webServices, config)
+	if err != nil {
+		return nil, err
+	}
+	// Remove any non-API endpoints from aggregator's spec. aggregatorOpenAPISpec
+	// is the source of truth for all non-api endpoints.
+	aggregator.FilterSpecByPaths(s.aggregatorOpenAPISpec, []string{"/apis/"})
+
+	// Build initial spec to serve.
+	specToServe, err := s.buildOpenAPISpec()
+	if err != nil {
+		return nil, err
+	}
+
+	// Install handler
+	s.openAPIService, err = handler.RegisterOpenAPIService(
+		specToServe, "/swagger.json", pathHandler)
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// openAPISpecInfo is used to store OpenAPI spec with its priority.
+// It can be used to sort specs with their priorities.
+type openAPISpecInfo struct {
+	serviceName string
+	spec        *spec.Swagger
+	priority    int32
+}
+
+// byPriority can be used in sort.Sort to sort specs with their priorities.
+type byPriority []openAPISpecInfo
+
+func (a byPriority) Len() int      { return len(a) }
+func (a byPriority) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a byPriority) Less(i, j int) bool {
+	if a[i].priority != a[j].priority {
+		// Sort by priority, higher first
+		return a[i].priority > a[j].priority
+	} else {
+		// Sort by service name.
+		return a[i].serviceName < a[j].serviceName
+	}
+}
+
+// buildOpenAPISpec aggregates all OpenAPI specs.  It is not thread-safe.
+func (s *openAPIAggregator) buildOpenAPISpec() (specToReturn *spec.Swagger, err error) {
+	specToReturn, err = aggregator.CloneSpec(s.localDelegatesOpenAPISpec)
+	if err != nil {
+		return nil, err
+	}
+	if err := aggregator.MergeSpecs(specToReturn, s.aggregatorOpenAPISpec); err != nil {
+		return nil, fmt.Errorf("cannot merge local delegate spec with aggregator spec: %s", err.Error())
+	}
+	specs := []openAPISpecInfo{}
+	for name, specInfo := range s.openAPISpecs {
+		specs = append(specs, openAPISpecInfo{name, specInfo.spec, specInfo.priority})
+	}
+	sort.Sort(byPriority(specs))
+	for _, specInfo := range specs {
+		if err := aggregator.MergeSpecs(specToReturn, specInfo.spec); err != nil {
+			return nil, err
+		}
+	}
+	return specToReturn, nil
+}
+
+// updateOpenAPISpec aggregates all OpenAPI specs.  It is not thread-safe.
+func (s *openAPIAggregator) updateOpenAPISpec() error {
+	if s.openAPIService == nil {
+		return nil
+	}
+	specToServe, err := s.buildOpenAPISpec()
+	if err != nil {
+		return err
+	}
+	return s.openAPIService.UpdateSpec(specToServe)
+}
+
+// inMemoryResponseWriter is a http.Writer that keep the response in memory.
+type inMemoryResponseWriter struct {
+	header   http.Header
+	respCode int
+	data     []byte
+}
+
+func (r *inMemoryResponseWriter) Header() http.Header {
+	if r.header == nil {
```

Comment:
Do we need to worry about access contention here? Would we better off just using a constructor which insures this exists?